src/Avalonia.Base/Data/Core/ExpressionNodes/ExpressionNode.cs

```diff
@@ -101,28 +92,26 @@ public void SetOwner(BindingExpression owner, int index)
     /// </param>
     public void SetSource(object? source, Exception? dataValidationError)
     {
-        var oldSource = Source;
-
-        if (source == AvaloniaProperty.UnsetValue)
-            source = null;
+        if (_source?.TryGetTarget(out var oldSource) != true)
+            oldSource = AvaloniaProperty.UnsetValue;
 
         if (source == oldSource)
             return;
 
-        if (oldSource is not null)
+        if (oldSource is not null && oldSource != AvaloniaProperty.UnsetValue)
             Unsubscribe(oldSource);
 
-        _source = new(source);
-
-        if (source is null)
+        if (source == AvaloniaProperty.UnsetValue)
         {
-            // If the source is null then the value is null. We explicitly do not want to call
+            // If the source is unset then the value is null. We explicitly do not want to call
             // OnSourceChanged as we don't want to raise errors for subsequent nodes in the
             // binding change.
+            _source = null;
             _value = AvaloniaProperty.UnsetValue;
```

Comment:
The comment states that "the value is null" but it's set to (unset) here. Should it be "the value is unset"?

Reply:
Good catch! fixed.